### PR TITLE
[Performance][MoE] Use bf16 router GEMM with fp32 accumulation for bf16-origin weights

### PR DESCRIPTION
**Pull request by Levi-JQ**
## What this PR does / why we need it

`AscendGateLinear` (PR #9617) forces fp32 weights and fp32 computation for the MoE router gate, because the previous `bf16×bf16→bf16` path (bf16 accumulation, no fp32 output) caused expert-selection errors in agent workloads on NPU.

However, the fp32 path (`aclnnMatmulV2`) is ~6x slower than necessary for bf16-origin weights — the common case for models like Kimi K3, whose gate weights are stored as bf16 in safetensors.

This PR adds a bf16 fast path: when the checkpoint weight is detected as bfloat16 at load time (via `weight_loader`, before `copy_` upcasts to fp32), the router GEMM runs as:

```python
torch.mm(x_bf16, w_bf16.t(), out_dtype=torch.float32)
```

which dispatches to `aclnnMm MatMulV3` — bf16 multiplication with **fp32 accumulation**. For bf16-origin operands this is numerically equivalent to the fp32 path (each `bf16×bf16` product fits exactly in the fp32 accumulator; the only difference is K-dim summation order, ~1e-5).

### Key design decisions

- **No environment variable.** The fast path is enabled automatically when:
  - Checkpoint weight dtype is bfloat16 (recorded by overriding `weight_loader` — the only reliable observation point before `copy_` and NZ-repack)
  - Input dtype is bfloat16 (checked at `forward` time, no risky cast)
  - Output dtype is fp32 (so `torch.mm(out_dtype=fp32)` is the only out_dtype supported by `aclnnMm MatMulV3` on NPU)
  - No bias (fast path does not handle bias)
  
  All four conditions must hold simultaneously; otherwise the fp32 fallback is used.

- **No dtype conversion in `forward`.** The `_forward_bf16_router` method does not cast `x` or `weight` — it requires `x.dtype == bfloat16` as a precondition (checked by `_use_bf16_fast_path(x)`). This avoids silently introducing precision loss if upstream ever sends non-bf16-origin activations.

- **ND bf16 weight cache (measured).** The bf16 cache stays in ND layout: a layout micro-benchmark (M=3/24) showed bit-identical outputs for ND vs FRACTAL_NZ caches, no measurable benefit from NZ on this `aclnnMm MatMulV3` path, and a pre-transposed-NZ cache ~40% slower (20.1/21.8µs vs 14.7/14.2µs batched). `.t()` on ND is a free stride view (aclnnMm handles transB natively).

- **Cache invalidation on weight reload.** `weight_loader` resets `_bf16_weight` so RL weight-update flows (which reload weights in-place) are reflected on the next forward.

- **Device-mismatch fix.** The `_bf16_weight` cache is rebuilt when `self.weight.device` changes, fixing the issue raised in the original PR review.

- **Forward compatibility.** Models with true fp32 router weights, non-bf16 input, non-fp32 output dtype, or bias are unaffected — the fp32 fallback path is preserved verbatim.

## Background: why the old path was slow

PR #9617 forced fp32 because the pre-existing path was `bf16×bf16→bf16` (bf16 accumulation). The fast path uses `bf16×bf16→fp32` (fp32 accumulation via `out_dtype=torch.float32`), which avoids the precision issue while being ~6x faster:

| Path | Multiply | Accumulate | Output | NPU kernel | Speed |
|------|----------|------------|--------|------------|-------|
| Pre-#9617 (bf16 F.linear) | bf16 | bf16 | bf16 | aclnnMatmulV2 | baseline (broken) |
| #9617 (fp32 F.linear) | fp32 | fp32 | fp32 | aclnnMatmulV2 | 84.8µs |
| **This PR (bf16 mm + fp32 out)** | **bf16** | **fp32** | **fp32** | **aclnnMm MatMulV3** | **13.7µs (6.2x)** |

## Precision verification

**Methodology**: the fp32 path (`F.linear` → aclnnMatmulV2) and the bf16 path (`torch.mm(out_dtype=fp32)` → aclnnMm MatMulV3) both run **on NPU** with the production ops; the float64 reference runs **on CPU**. Shape: N=896, K=7168 (Kimi K3 router).

### M-sweep vs the float64 reference (16 shapes, M = 1 to 896)

Fixed bf16-origin weight [896, 7168]; 16 M values uniformly spaced over [1, 896]; relative-error denominator = |y_fp64| clamped at 1e-06 (near-zero logits excluded). In both tables, **ratio = bf16 error / fp32 error** (each measured against the same float64 reference); ratio < 1 means the bf16 path is closer to the reference. Top-16 expert selection (N=896, k=16, per-row expert-set match) is measured on the same outputs and reported alongside the absolute-error table.

#### Absolute error

| M | fp32 mean_abs | fp32 max_abs | bf16 mean_abs | bf16 max_abs | ratio (mean) | top16 (bf16~fp32) |
|---:|---:|---:|---:|---:|---:|---:|
| 1 | 4.3376e-05 | 6.6163e-03 | 2.2527e-05 | 1.1962e-03 | 0.52 | 1/1 |
| 61 | 4.3490e-05 | 1.5986e-02 | 2.4576e-05 | 5.6703e-03 | 0.57 | 61/61 |
| 120 | 4.4128e-05 | 1.5986e-02 | 2.4397e-05 | 5.6703e-03 | 0.55 | 120/120 |
| 180 | 4.3756e-05 | 1.5986e-02 | 2.4412e-05 | 6.1590e-03 | 0.56 | 180/180 |
| 240 | 4.3933e-05 | 1.6362e-02 | 2.4349e-05 | 6.1590e-03 | 0.55 | 240/240 |
| 299 | 4.3902e-05 | 1.6547e-02 | 2.4339e-05 | 6.1590e-03 | 0.55 | 299/299 |
| 359 | 4.3729e-05 | 1.7523e-02 | 2.4393e-05 | 7.3646e-03 | 0.56 | 359/359 |
| 419 | 4.3826e-05 | 1.7976e-02 | 2.4368e-05 | 7.3646e-03 | 0.56 | 419/419 |
| 478 | 4.3911e-05 | 1.8174e-02 | 2.4373e-05 | 7.3646e-03 | 0.56 | 478/478 |
| 538 | 4.3935e-05 | 1.8174e-02 | 2.4327e-05 | 7.3646e-03 | 0.55 | 538/538 |
| 598 | 4.4046e-05 | 1.9968e-02 | 2.4318e-05 | 7.3646e-03 | 0.55 | 598/598 |
| 657 | 4.3770e-05 | 1.6547e-02 | 2.4355e-05 | 7.3646e-03 | 0.56 | 657/657 |
| 717 | 4.3955e-05 | 1.8466e-02 | 2.4327e-05 | 7.3646e-03 | 0.55 | 717/717 |
| 777 | 4.3988e-05 | 2.0267e-02 | 2.4338e-05 | 7.3646e-03 | 0.55 | 777/777 |
| 836 | 4.4007e-05 | 2.0267e-02 | 2.4338e-05 | 7.3646e-03 | 0.55 | 836/836 |
| 896 | 4.4007e-05 | 2.0267e-02 | 2.4338e-05 | 7.3646e-03 | 0.55 | 896/896 |

#### Relative error

| M | fp32 mean_rel | fp32 max_rel | bf16 mean_rel | bf16 max_rel | ratio (mean) |
|---:|---:|---:|---:|---:|---:|
| 1 | 1.1324e-06 | 6.9017e-05 | 7.7533e-07 | 3.8605e-05 | 0.68 |
| 61 | 2.1194e-06 | 1.0449e-02 | 1.3590e-06 | 4.0311e-03 | 0.64 |
| 120 | 2.8259e-06 | 5.3189e-02 | 1.5852e-06 | 1.2173e-02 | 0.56 |
| 180 | 2.3724e-06 | 1.5573e-02 | 1.4727e-06 | 1.2173e-02 | 0.62 |
| 240 | 2.5997e-06 | 5.3189e-02 | 1.4884e-06 | 1.2173e-02 | 0.57 |
| 299 | 2.6227e-06 | 5.3189e-02 | 1.4894e-06 | 1.2173e-02 | 0.57 |
| 359 | 2.4289e-06 | 2.1184e-02 | 1.4652e-06 | 1.2173e-02 | 0.60 |
| 419 | 2.5630e-06 | 5.3189e-02 | 1.5287e-06 | 1.2173e-02 | 0.60 |
| 478 | 2.4593e-06 | 2.9177e-02 | 1.5161e-06 | 1.2173e-02 | 0.62 |
| 538 | 2.6285e-06 | 5.3189e-02 | 1.5371e-06 | 1.2173e-02 | 0.58 |
| 598 | 2.5640e-06 | 5.3189e-02 | 1.5338e-06 | 1.2173e-02 | 0.60 |
| 657 | 2.6290e-06 | 5.3189e-02 | 1.5390e-06 | 1.2173e-02 | 0.59 |
| 717 | 2.6185e-06 | 5.3189e-02 | 1.5392e-06 | 1.2173e-02 | 0.59 |
| 777 | 2.5851e-06 | 5.3189e-02 | 1.5225e-06 | 1.2173e-02 | 0.59 |
| 836 | 2.6553e-06 | 5.3189e-02 | 1.5480e-06 | 1.2173e-02 | 0.58 |
| 896 | 2.6464e-06 | 5.3189e-02 | 1.5573e-06 | 1.2173e-02 | 0.59 |

Findings:

- The bf16 path is closer to the float64 reference on **16/16 shapes** in both metrics: absolute-error ratio stable at **0.52–0.57** (avg 0.55), relative-error ratio at **0.56–0.68** (avg ~0.60).
- mean_abs is essentially flat across M (fp32: 4.34–4.41e-05, bf16: 2.25–2.46e-05) — the error is set by the K=7168 fp32-accumulation structure, not by M; max_abs grows slowly with M purely by extreme-value statistics.
- mean_rel sits at the ~1e-06 scale; max_rel is dominated by near-zero logits and saturates.
- **Top-16 expert selection is 100% consistent at every shape in this sweep**: bf16 vs fp32 matches on **7176/7176** rows across all 16 shapes, and both paths vs the float64 reference are likewise 100% (7176/7176 each).

## Does this PR introduce any user-facing change?

No. The fast path is transparent — it activates automatically for bf16-origin weights and falls back to the existing fp32 path otherwise.

## How was this patch tested

- Precision test on NPU (Ascend910): compared fp32 path, bf16 path, and float64 CPU reference across M=3/24/961 with bf16-origin weights (a true-fp32-weight control, forced through the bf16 math, degrades to 66.7% top-16 consistency — confirming the gate is required and works)
- M-sweep (16 shapes, M=1..896): absolute / relative errors vs float64 CPU reference — bf16 closer on 16/16 shapes; top-16 expert-set match 100% at every shape (bf16 vs fp32: 7176/7176 rows; both paths vs float64 likewise)
- Layout micro-benchmark (nz_bench.py): ND / FRACTAL_NZ / pre-transposed-NZ bf16 caches produce bit-identical outputs; ND + `.t()` selected (pre-transposed NZ ~40% slower)
- `weight_loader` override correctly records checkpoint dtype from safetensors header and invalidates the bf16 cache on reload
- NZ format applied to fp32 weights by `process_weights_after_loading` (unchanged); bf16 cache deliberately kept in ND per the layout benchmark above
- vLLM main: https://github.com/vllm-project/vllm/commit/84030bbe3d74d99bad477a3d2e37a973ccd8865c

### CODE REVIEW

File: vllm_ascend/ops/fused_moe/gate_linear.py [hunk 1]
```diff
@@ -51,8 +69,65 @@ def __init__(
...
+    def weight_loader(
+        self, param: Parameter, loaded_weight: torch.Tensor
+    ):
+        # Record the original checkpoint dtype before copy_ upcasts to fp32.
+        # This is the only reliable observation point: after copy_ the
+        # Parameter is fp32, and after process_weights_after_loading the
+        # data may be NZ-repacked making bit-level inspection unreliable.
+        if self._checkpoint_weight_dtype is None:
+            self._checkpoint_weight_dtype = loaded_weight.dtype
+        super().weight_loader(param, loaded_weight)
```
[the remark this answers is not in the document]
Comment (Levi-JQ) [reply vllm_ascend/ops/fused_moe/gate_linear.py]: done

**Comment by weijinqian0** [comment]
Need check deepseek v4. This Pr may lead to precision problem.

**Comment by Levi-JQ** [comment]
> Need check deepseek v4. This Pr may lead to precision problem.

DSV4 is not affected by this PR. Verified against the actual DeepSeek-V4-Flash checkpoint:

Different code path: DSV4's gate is a ReplicatedLinear, not GateLinear — it never touches AscendGateLinear, the only class this PR modifies. Its forward explicitly computes F.linear(hidden_states.float(), self.gate.weight) with precast_fp32_weight = True, so the router stays on the fp32 MatMulV2 path by design ((https://github.com/vllm-project/vllm-ascend/blob/8e74e741e8ef407c9d1f61ce26f504e06c983752/vllm_ascend/models/deepseek_v4/model.py#L292-L295)).

Double-guarded: even if DSV4 were switched to GateLinear in the future, the fast path requires all four conditions — checkpoint weight bf16, activation x.dtype == bf16, out_dtype == fp32, and no bias. DSV4 passes fp32 activations (https://github.com/vllm-project/vllm-ascend/blob/8e74e741e8ef407c9d1f61ce26f504e06c983752/vllm_ascend/models/deepseek_v4/model.py#L404-L405), so the gate condition fails and it falls back to the fp32 path.

The bf16 fast path only activates when both inputs are bf16-origin (lossless upcast, verified bit-exact top-16 on real Kimi K3 weights). Models that deliberately force fp32 — like DSV4 — are structurally excluded.
